### PR TITLE
refactor disease model components accommodate inheriting from Component

### PR DESCRIPTION
**Pull request by rmudambi**
### Description

- *Category*: refactor
- *JIRA issue*: [MIC-4473](https://jira.ihme.washington.edu/browse/MIC-4473)

### Changes and notes
- Simplifies `AgeOutSimulants`' time-step cleanup method - this is changed in this PR because the test started failing
- Refactors `DiseaseModel` and the various state and transition classes to accommodate the fact that their super-classes now inherit from `Component`. 
- Adds some typing where it wasn't increasing the scope of the PR too much
- Replaced previous `state.add_transition()` functions with `state.add_rate_transition()`, `state.add_proportion_transition()`, and `state.add_dwell_time_transition` which I think are clearer, and bring us in line with the new signature of `add_transition` in `State`. This also means we no longer have sub-classes having a method with the same name, but incompatible arguments as their super-class.

### Testing
All unit tests pass.

### CODE REVIEW

File: src/vivarium_public_health/disease/special_disease.py [hunk 1]
```diff
@@ -322,6 +346,7 @@ def continuous_filter(index):
 
         return filter_function
 
+    # todo: reorder these methods to make it easier to understand what they are each doing
```
Comment (patricktnast) [inline src/vivarium_public_health/disease/special_disease.py:349]: I think now would be a good time to address this todo.

File: src/vivarium_public_health/disease/state.py [hunk 2]
```diff
@@ -102,110 +120,124 @@ def _transition_side_effect(self, index, event_time):
...
+        """
+        if "proportion" not in get_data_functions:
+            raise ValueError("You must supply a proportion function.")
+
+        transition = ProportionTransition(self, output, get_data_functions, **kwargs)
+        self.add_transition(transition)
+        return transition
+
+
+# TODO is there a better name for the union of susceptible and recovered?
```
Comment (patricktnast) [inline src/vivarium_public_health/disease/state.py:182]: ChatIHME suggests "non-diseased" is an epidemiological term of art for this, and I think it is correct based on a quick follow-up

**Review by patricktnast — APPROVED** [review]

File: src/vivarium_public_health/disease/model.py [hunk 3]
```diff
@@ -56,7 +59,36 @@ def transition_names(self) -> List[TransitionString]:
             transitions += state.get_transition_names()
         return transitions
 
-    def setup(self, builder):
+    # @property
```
[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium_public_health/disease/model.py:62]: Almost. I need to flatten the commented out version. Thanks for catching.

**Comment by mattkappel** [comment]
Looks like this needs linting

**Review by mattkappel — APPROVED** [review]

**Pull request merged**